connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala

```diff
@@ -288,7 +289,21 @@ private[sql] class ProtobufDeserializer(
       var skipRow = false
       while (i < validFieldIndexes.length && !skipRow) {
         val field = validFieldIndexes(i)
-        val value = if (field.isRepeated || field.hasDefaultValue || record.hasField(field)) {
+
+        // If `materializeZeroValues` is true, the written field will contain the
+        // default zero value for its type (e.g. 0 for int, "" for string, etc:
+        // https://protobuf.dev/programming-guides/proto3/#default).
+        // Otherwise, the field will be null unless the serialized proto explicitly
+        // contains the field, or it has an explicit default (
+        // proto2 only https://protobuf.dev/programming-guides/proto2/#optional)
+        // Note that in proto3, the serialized proto will not contain the field
+        // if the value was explicitly set to its zero value. See:
+        // https://protobuf.dev/programming-guides/field_presence/#presence-in-proto3-apis
```

Comment:
Hmm, is it a possible bug? If a field is explicitly set to zero value, the serialized proto message won't contain it, so when deserializing the message we will get null instead of zero?

Reply:
@viirya I don't think this is a bug, according to proto3 default value spec,

> When a message is parsed, if the encoded message does not contain a particular singular element, the corresponding field in the parsed object is set to the default value for that field. These defaults are type-specific:
> 
> For strings, the default value is the empty string.
> For bytes, the default value is empty bytes.
> For bools, the default value is false.
> For numeric types, the default value is zero.
> For [enums](https://protobuf.dev/programming-guides/proto3/#enum), the default value is the first defined enum value, which must be 0.
> For message fields, the field is not set. Its exact value is language-dependent. See the [generated code guide](https://protobuf.dev/reference/) for details.
> 

What that mean is the absent optional field's value should be the default (0 for int, for example) if they are retrieved by the client -- this is very confusing (I know), it basically says there is by default (not using oneof keyword) there is no way to distinguish between absent field and fields populated with default value (like int64 field set to 0). Protobuf 3.15 somewhat bring this back, more details could be found [here](https://stackoverflow.com/questions/42622015/how-to-define-an-optional-field-in-protobuf-3)

Reply:
Oh, I meant the original behavior of protobuf deseralizing here. Before this PR, based on it describes, a field set with zero value will not be present in serialized message, so when deserializing it, it will be given a null instead of zero value originally?

Reply:
> Hmm, is it a possible bug? If a field is explicitly set to zero value, the serialized proto message won't contain it, so when deserializing the message we will get null instead of zero?
>  a field set with zero value will not be present in serialized message, so when deserializing it, it will be given a null instead of zero value originally?

@viirya i'm not sure; in some sense it could be considered a bug and in another sense it could be considered a convention. I think the behavior of filling in defaults makes sense because of the proto3 spec Pang linked above.

But because there is no way to distinguish between "field is set to zero value" and "field is not present" in proto3 serialized messages, deserialization libraries much choose what to do, either

a) if a field is not present in serialized value, deserialize as null [the default behavior]
- in case the field was not set originally, this is "correct"
- in case the field was set to the zero value this is "incorrect"

b) if a field is not present, deserialize to its zero value [what i tried to add in this PR]
- in case the field was not set originally, this is "incorrect"
- in case the field was set to the zero value, this is "correct"

So either way the library chooses, one of the cases is kind of strange (i.e. you set a value and it doesn't come out, or you didn't set a value and yet a default comes out). Given this, it does feel like it's not clear which is the right "default". It also depends on whether the original message is proto2 or proto3, i think. Maybe what we would actually want is:

- for proto3 this zero-value materialization to be the default behavior
- for proto2 we'd want to look for field presence

Reply:
One more think to add is @justaparth, with this PR in, proto2 and proto3.15 message with optional key word will still work as expected? So nothing will be break.

Reply:
@pang-wu thanks for the callout, i'll double check that we're testing those cases and if not add a new one.

In the meantime, i've refactored the code here slightly. I _think_ it might be simpler/correct to define the behavior through the lens of [field presence](https://protobuf.dev/programming-guides/field_presence/), i.e.
- if a field has presence info, we will deserialize it if present and otherwise fill in `null`
- if a field **doesn't** have presence info, we only deserialize it if it is explicitly set, or is proto2 with an explicit default.

And then, on top of that logic (which mirrored the logic as-it-was), i've added the `materializeZeroValues` option which will materialize the default if a field does not have presence information available.

Does this make more sense?

I think now it's easier to see how the current code is treating fields without presence info, i.e. it's choosing to deserialize to null instead of choosing a default. If we think that's the right behavior going forward, maybe we could consider changing the default behavior of the code (though i'm afraid to break any existing applications that may use this).

Reply:
hey @viirya just following up if we could get another review on this? thank you!

Reply:
I've done a quick update to this PR! I've added some additional tests and made the code more clear as to how field presence is the main factor to consider when deserializing

Reply:
ok quick update, i took another pass at this based on some comments below and i think the code is much clearer in its intent now!